docroot/modules/humsci/hs_capx/src/Capx.php

```diff
@@ -0,0 +1,230 @@
+
+/**
+ * Service to test CAPx connection and import organization data.
+ */
+class Capx {
+
+  /**
+   * API url used for organization data.
+   */
+  const API_URL = 'https://api.stanford.edu';
+
+  /**
+   * Authentication url.
+   */
+  const AUTH_URL = 'https://authz.stanford.edu/oauth/token';
+
+  /**
+   * The actual CAP API.
+   */
+  const CAP_URL = 'https://cap.stanford.edu/cap-api/api/profiles/v1';
+
+  /**
+   * CAPx API username.
+   *
+   * @var string
+   */
+  protected $username;
+
+  /**
+   * CAPx API password.
+   *
+   * @var string
+   */
+  protected $password;
+
+  /**
+   * Capx constructor.
+   *
+   * @param \GuzzleHttp\ClientInterface $guzzle
+   *   Guzzle Client service.
+   * @param \Drupal\Core\Cache\CacheBackendInterface $cache
+   *   Cache service.
+   * @param \Drupal\Core\Database\Connection $database
+   *   Database connection service.
+   */
+  public function __construct(ClientInterface $guzzle, CacheBackendInterface $cache, Connection $database) {
+    $this->client = $guzzle;
+    $this->cache = $cache;
+    $this->database = $database;
+  }
+
+  /**
+   * Set the CAPx Username.
+   *
+   * @param string $username
+   *   Username.
+   */
+  public function setUsername($username) {
+    $this->username = $username;
+  }
+
+  /**
+   * Set the CAPx Password.
+   *
+   * @param string $password
+   *   Password.
+   */
+  public function setPassword($password) {
+    $this->password = $password;
+  }
+
+  /**
+   * Get the url for CAPx for the given organizations.
+   *
+   * @param string $organizations
+   *   Comma separated organization codes.
+   * @param bool $children
+   *   Include all children of the organizations.
+   *
+   * @return string
+   *   CAPx URL.
+   */
+  public static function getOrganizationUrl($organizations, $children = FALSE) {
+    $organizations = preg_replace('/[^A-Z,]/', '', strtoupper($organizations));
+    $url = self::CAP_URL . "?orgCodes=$organizations&ps=1000";
+    if ($children) {
+      $url .= '&includeChildren=true';
+    }
+    return $url;
+  }
+
+  /**
+   * Get the url for CAPx for given workgroups.
+   *
+   * @param string $workgroups
+   *   Commas separated list of workgroups.
+   *
+   * @return string
+   *   CAPx URL.
+   */
+  public static function getWorkgroupUrl($workgroups) {
+    $workgroups = preg_replace('/[^A-Z,:]/', '', strtoupper($workgroups));
+    return self::CAP_URL . "?privGroups=$workgroups&ps=1000";
+  }
+
+  /**
+   * Test the connection with the username and passwords is valid.
+   *
+   * @return bool
+   *   The connection was successful.
+   */
+  public function testConnection() {
+    $parameters = ['grant_type' => 'client_credentials'];
+
+    try {
+      $response = $this->client->request('GET', self::AUTH_URL, [
+        'query' => $parameters,
+        'auth' => [$this->username, $this->password],
+      ]);
+    }
+    catch (GuzzleException $e) {
+      // Most errors originate from the API itself.
+      return FALSE;
```

Comment:
Should log em. Debugging these things is hard.

Reply:
Logging added